src/compiler/transformers/esnext.ts

```diff
@@ -118,28 +118,46 @@ namespace ts {
         }
 
         function visitYieldExpression(node: YieldExpression) {
-            if (enclosingFunctionFlags & FunctionFlags.Async && enclosingFunctionFlags & FunctionFlags.Generator && node.asteriskToken) {
-                const expression = visitNode(node.expression, visitor, isExpression);
-                return setOriginalNode(
-                    setTextRange(
-                        createYield(
-                            createAwaitHelper(context,
-                                updateYield(
-                                    node,
-                                    node.asteriskToken,
-                                    createAsyncDelegatorHelper(
-                                        context,
-                                        createAsyncValuesHelper(context, expression, expression),
-                                        expression
+            if (enclosingFunctionFlags & FunctionFlags.Async && enclosingFunctionFlags & FunctionFlags.Generator) {
+                if (node.asteriskToken) {
+                    const expression = visitNode(node.expression, visitor, isExpression);
+
+                    return setOriginalNode(
+                        setTextRange(
+                            createYield(
+                                createAwaitHelper(context,
+                                    updateYield(
+                                        node,
+                                        node.asteriskToken,
+                                        createAsyncDelegatorHelper(
+                                            context,
+                                            createAsyncValuesHelper(context, expression, expression),
+                                            expression
+                                        )
                                     )
                                 )
-                            )
+                            ),
+                            node
                         ),
                         node
-                    ),
-                    node
-                );
+                    );
+                }
+
+                if (node.expression && node.expression.kind !== SyntaxKind.AwaitExpression) {

```

Comment:
The actual ECMAScript specs don't care whether you are already awaiting the operand to `yield`, e.g. in `yield await x`, you `await` _twice_: once for the `await` and once for the `yield`. It also doesn't matter if `yield` has an expression. Rather, `yield` is treated as `yield undefined`, and as a result we still have to await `undefined` before we actually yield.

It seems inefficient, but the runtime semantics would be observably different. If we don't have the same number of underling [Awaits](https://tc39.github.io/ecma262/#await) as the spec, then you could end up with observable differences in resolution timing between our downlevel emit and uplevel ECMAScript runtime behavior for the same code.

@DanielRosenwasser, @bterlson: Any thoughts?

Reply:
@agentcooper: in practice this would mean replacing line 151 with `node.expression ? visitNode(node.expression, visitor, isExpression) : createVoidZero()` and removing the `if` condition on line 146.

Reply:
@rbuckton got it. should I go ahead with the change or are we waiting for others to reply?

Reply:
Yes, go ahead and make the change.